### PR TITLE
feat: Custom request types, websocket rewrite, client custom responses

### PR DESCRIPTION
**Pull request by 0x00002a**
So this one is a bit of a mess. The websocket rewrite should probably have been its own PR but I didn't realise that until I was almost done with it. Anyway, deep breath:

--- 

### Core 
- Filter concept introduced, the client and server each have speciailised versions of this
- file filter added as a utility filter for reading or writing to/from a file with a request/response. Its more an example for now
- some beast and asio stuff has been forwarded to the `malloy` namespace. `error.hpp` now holds a forwarded `malloy::error_code`

### Client 
- Filter concept which requires:
    - request_type
    - header_type
    - `auto body_for(const header_type&) -> std::variant<...>`
    - `void setup_body(const header_type&, T&)` where T is one of the types in the variant returned by `body_for` (its a visitor)
    - std::move_constructible
- http(s)_request now 
    - Uses an async handler function which models a visitor over the return type of `Filter::body_for`
    - Has an optional `Filter` param to provide a custom filter for the response
    - Returns std::future<malloy::error_code> instead of a request. It has an async handler
- The async handler is a visitor over `response<T>...` where `T` is the bodies returned by `setup_body` for the filter (this means that with the default handler, it will always be `response<string_body>`)

### Server
- router::add now takes an optional argument that allows setting a request filter. Request filters are simpler than response filters based on the assumption that the server will expect 1 and only 1 type of body. This could be changed 
- request_filter concept which requires
    - request_type (which must also be a single request type, NOT a variant) 
    - `setup_body(const request_type::header_type&, request_type::body_type::value_type&)` 
    - std::move_constructible

### Websockets
The websocket API has been redesigned. The server and client parts have been merged into a single class which uses a `stream` type that does the handling of tls vs plain (although I don't think websocket tls is implemented yet?). The interface is based around providing a connection object which represents the websocket session and can be used to manage said session. The API is entirely async (apart from stop). It uses callbacks although really I would like to use coroutines here (but I lack experience with them). 

#### Server
 On the server side, the new interface passes the connection object as a `shared_ptr` directly to the handler, along with the request (as a `string_body`) that kicked it off, it is then up to the handler whether to call `accept` on the connection object.

#### Client 
The client side is similar except its callback takes in `error_code` instead of the request. If the error code is truthy then the connection object will be nullptr, else its safe to use. The handshake is done before the handler is invoked so the handler should call either `read` or `send`.

---

I've added examples for the new filters and fixed the examples for the websockets so check those out for, well, examples. I will sort out the merge conflicts if/when this is ready for merging, as I think its going to require quite a bit of manual sorting out. I don't think its ready for merge currently, I want feedback on it and it could do with some more testing. Also needs some documentation, I haven't updated the doxygen comments or added new ones for the most part.

Closes: #17

### CODE REVIEW

**Comment by 0x00002a** [comment]
Just rebasing this now, so don't pull it :p

**Comment by 0x00002a** [comment]
Alright nvm, trying to rebase this is a nightmare. I just merged the upstream changes, I guess its ok anyway since it will be rebased on merge?

**Comment by Tectu** [comment]
Wow.
I eeeh... will need some time to process this PR :-D

**Comment by Tectu** [comment]
I started reviewing this. So far, most seems to be related to coding style, can you modify this PR accordingly?

coding style fixes (I'll update `coding_style.md` accordingly):
- No empty lines at the beginning of a file
- Empty line at the end of a file
- Opening curly-braces of a function body should be on a new line

**Comment by Tectu** [comment]
It would also be great if you could add some Doxygen to document the new classes/interfaces. Doesn't need to be complete at the moment - I'll put some effort into documentation once `v0.1` is ready to be released. But at least class/interface documentation would be great :)

**Comment by Tectu** [comment]
Did you by any chance miss my other post on this PR regarding the coding style changes? :p

**Comment by 0x00002a** [comment]
Ah, yes I did 🤦. I would prefer to do it via clang-format, but that will affect everything not just changes from this PR. I could localise it to only the _files_ affected by the PR though (although I think thats most of them)

**Comment by Tectu** [comment]
Works for me. Alternatively we can also ignore the coding style for this PR (as an exception - because it's you 😛), I will finish the review process ignoring coding style and after that I fix/improve the `.clang-format` and we add this to the workflow.

What do you think?

**Comment by 0x00002a** [comment]
This PR is already quite messy, it would be cleaner to merge then run clang-format over the whole codebase imo, since not all of that follows the current standard (at least mine doesn't :p). Once the codebase is formatted according to the `.clang-format` file then we stop anything being merged that doesn't follow it, that way we keep the main clean and ensure formatting PRs doesn't cause a pile of conflicts.

**Comment by Tectu** [comment]
Agreed.

File: .gitignore [hunk 1]
```diff
@@ -91,4 +91,5 @@ modules.xml
 # Sonarlint plugin
 .idea/sonarlint
 
-# End of https://www.toptal.com/developers/gitignore/api/clion+all
\ No newline at end of file
+# End of https://www.toptal.com/developers/gitignore/api/clion+all
+/.vs
```
Comment (Tectu) [inline .gitignore:95]: This shouldn't be part of this PR as it seems unrelated to the subject of this PR.

File: examples/ws_handlers.hpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+#include <malloy/websocket/connection.hpp>
+
+#include <thread>
+
+namespace malloy::examples::ws {
+template<bool isClient>
+using ws_connection = std::shared_ptr<malloy::websocket::connection<isClient>>;
+
+
+template<bool isClient>
```
Comment (Tectu) [inline examples/ws_handlers.hpp:19]: This could use some short comment describing it's functionality.

File: examples/ws_handlers.hpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+void oneshot_read(const ws_connection<isClient>& conn, std::invocable<malloy::error_code, std::string> auto&& on_read) {
+    auto buffer = std::make_shared<boost::beast::flat_buffer>();
+    conn->read(*buffer, [buffer, on_read = std::forward<decltype(on_read)>(on_read)](auto ec, auto) {
+        on_read(ec, malloy::buffers_to_string(malloy::buffer(buffer->cdata(), buffer->size())));
+    });
+}
+
+template<bool isClient>
```
Comment (Tectu) [inline examples/ws_handlers.hpp:27]: This could use some short comment describing it's functionality.

File: examples/ws_handlers.hpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+void oneshot(const auto& req, ws_connection<isClient> conn, const std::string& msg_data) {
+    auto msg = std::make_shared<std::string>(msg_data); // Keep message memory alive
+    conn->accept(req, [msg, conn] { conn->send(malloy::buffer(*msg), [msg](auto, auto) {}); });
+}
+
+template<bool isClient>
```
Comment (Tectu) [inline examples/ws_handlers.hpp:33]: This could use some short comment describing it's functionality.

File: examples/ws_handlers.hpp [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    boost::beast::flat_buffer buff_;
+
+
+};
+using server_echo = ws_echo<false>;
+
+
+
+template<bool isClient>
+class ws_timer: public std::enable_shared_from_this<ws_timer<isClient>> {
```
Comment (Tectu) [inline examples/ws_handlers.hpp:78]: This could use some short comment describing it's functionality.

File: lib/malloy/client/controller.hpp [hunk 6]
```diff
@@ -19,6 +34,20 @@ namespace malloy::client
         class connection_plain;
     }
 
+    namespace detail {
+        
+        struct default_resp_filter {
```
Comment (Tectu) [inline lib/malloy/client/controller.hpp:39]: This could use some short comment describing it's functionality/characteristics/purpose/behavior.

File: lib/malloy/client/type_traits.hpp [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+#include "malloy/type_traits.hpp"
+
+namespace malloy::client::concepts {
+
+    template<typename F>
+    void filter_helper(const F& f) {
+    }
+
+template<typename F>
```
Comment (Tectu) [inline lib/malloy/client/type_traits.hpp:15]: Missing indentation (entire concept prototype & body).

File: lib/malloy/client/type_traits.hpp [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        std::visit([](auto&& v) {
+            F f2;
+            typename F::header_type h2;
+            typename std::decay_t<decltype(v)>::value_type r;
+            f2.setup_body(h2, r);
+            }, f.body_for(h)) 
+    };
+}; 
+
```
Comment (Tectu) [inline lib/malloy/client/type_traits.hpp:27]: Unnecessary empty line.

File: lib/malloy/error.hpp [hunk 9]
```diff
@@ -0,0 +1,9 @@
+
+#pragma once
+
+#include <boost/beast/core/error.hpp>
+
+namespace malloy {
+	using error_code = boost::beast::error_code;
+
```
Comment (Tectu) [inline lib/malloy/error.hpp:9]: Unnecessary empty line(s).

File: lib/malloy/http/filters/file.hpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+#include "malloy/http/response.hpp"
+#include "malloy/http/request.hpp"
+
+#include <filesystem>
+#include <variant>
+
+namespace malloy::http::filters {
+
+template<bool isRequest>
```
Comment (Tectu) [inline lib/malloy/http/filters/file.hpp:12]: Missing class/struct documentation.

File: lib/malloy/http/type_traits.hpp [hunk 11]
```diff
@@ -0,0 +1,11 @@
+
+#pragma once
+
+#include <boost/beast/http/type_traits.hpp>
+
+namespace malloy::http::concepts {
+
+template<typename B>
```
Comment (Tectu) [inline lib/malloy/http/type_traits.hpp:8]: Missing indentation.

File: lib/malloy/server/routing/endpoint_http_files.hpp [hunk 12]
```diff
@@ -1,39 +1,46 @@
...
+#include "malloy/http/uri.hpp"
 
 #include <filesystem>
 
 namespace malloy::server
 {
-
-    struct endpoint_http_files :
-        endpoint_http
+    class endpoint_http_files :
```
Comment (Tectu) [inline lib/malloy/server/routing/endpoint_http_files.hpp:11]: Missing class documentation.

File: test/test_suites/components/uri.cpp [hunk 13]
```diff
@@ -2,6 +2,8 @@
 
 #include <malloy/http/uri.hpp>
 
+#include <iostream>
```
Comment (Tectu) [inline test/test_suites/components/uri.cpp:5]: Is this really necessary? This include was added to `test.hpp`: https://github.com/Tectu/malloy/blob/a20816b1488824bf185c5a77139c3bbec290b703/test/test.hpp#L12

File: test/test_suites/components/websockets.cpp [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+				}
+			);
+			});
+
+		s_ctrl.start();
+		c_ctrl.start();
+
+		stop_token.wait();
+		CHECK(server_recieved);
+
```
Comment (Tectu) [inline test/test_suites/components/websockets.cpp:87]: Lots of empty lines.

**Review by Tectu — CHANGES_REQUESTED** [review]

File: examples/ws_handlers.hpp (hunk 4 above)
Comment (0x00002a) [reply examples/ws_handlers.hpp:33]: I think it should really be named `oneshot_send` actually, I'll rename it

File: lib/malloy/client/controller.hpp (hunk 6 above)
Comment (0x00002a) [reply lib/malloy/client/controller.hpp:39]: Is there an `internal` group or something I can put it into? Since its meant to be part of the API (but has to be because header files)

Comment (Tectu) [reply lib/malloy/client/controller.hpp:39]: I was wondering the same thing :p
Currently there is no such namespace but personally I think that `detail` might be a bit inaccurate here.

If a user is not supposed to ever use this directly I think that `internal` is a better namespace for this.

Comment (0x00002a) [reply lib/malloy/client/controller.hpp:39]: I was thinking about doxygen :p, but in terms of namespacing I went with `detail` because thats what most libraries I've come across use. I can change it to `internal` here but imo it would be better to do that across the whole codebase to maintain consistency (since I've used `detail` in #17 as well)

Comment (Tectu) [reply lib/malloy/client/controller.hpp:39]: Agreed, let's keep `detail`.

**Comment by 0x00002a** [comment]
`oneshot` has been renamed to `accept_and_send`, I've removed `filter_helper` since its unused

**Review by Tectu — APPROVED** [review]
